### ISSUE TITLE
Rain doesn't reach the bottom of the screen

### ISSUE THREAD

**Issue by Pikamander2** [OpenRCT2/OpenRCT2#5402]
**OS:** Windows 10
**Version:** 2.01.028
**Commit/Build:** 0.0.7-develop-f26380c

The bottom of the screen doesn't show rain. It lines up perfectly with the money counter and calendar, so it seems like there's an invisible bar or something covering up the rain and causing it to not get drawn.

**Steps to reproduce:**
1. Start a new game
2. Open the cheats menu
3. Force the weather to be rain (or heavy rain, or thunderstorm)
4. The bottom of the screen will not have any rain

**Screenshots / Video:**

![image](https://cloud.githubusercontent.com/assets/5246872/25422720/cfbaac54-2a2f-11e7-9bc9-5aa4a9915185.png)

![image](https://cloud.githubusercontent.com/assets/5246872/25422798/1ede374c-2a30-11e7-856e-31fd4551cdaa.png)

**Comment by Gymnasiast** [comment]
I was certain this was a duplicate, but I can't seem to find the bug, so I guess this wasn't reported because it was too obvious.

**Comment by janisozaur** [comment]
https://github.com/OpenRCT2/OpenRCT2/issues/2607

**Comment by Gymnasiast** [comment]
#2607 has nothing to do with this.

**Comment by janisozaur** [comment]
> Possibly related: Rain is not rendered in bottom and top areas (with toolbars)

**Comment by Gymnasiast** [comment]
" Possibly "
These are two different issues that _might_ be related, but still are two different issues.

**Comment by duncanspumpkin** [comment]
This is a duplicate it's all because it draws rain too late.

**Comment by IntelOrca** [comment]
I don't think it should be marked as duplicate, as @Gymnasiast said - issues should be based on issues rather than causes. Otherwise it is more likely to get raised again. A fix can fix one or more issues.

**Comment by MeesterDev** [comment]
It might be superfluous to add this, but the rain is also missing from the top of the screen (also visible in the screenshots). [I'm using 0.0.8-develop 787ddf8]

**Comment by Nubbie** [comment]
related to #676 & #2249

**Comment by suipaste** [comment]
As far as I can see there is a gap at the top and the bottom because the rain is drawn on top of the tool bars. I changed the height of them to see what happened and the rain drew on top.

I'm new to openrct2 development but does anyone know where the code which governs the draw order occurs?

**Comment by JoelLarson** [comment]
I've debugged this issue and think there are a couple of solutions that could work. I validated the rain can be drawn provided there isn't a window with a transparent background.

![image](https://user-images.githubusercontent.com/519553/104998079-dde4aa00-59e7-11eb-9ec9-7544450700c7.png)

The solutions seem to be:
* Changing the rain drawing to draw over the empty area of windows
* Create new "windows" for each group of widgets to remove the transparent background
* Rework rain drawing

I think the second solution is the most practical as the rain drawing is currently a recursive headache.

**Comment by duncanspumpkin** [comment]
What we want to do is make effects like this apply to the viewport post rendering but prior to ui drawing. We haven't split up the ui from the rest of the viewport drawing yet though.

**Comment by 733737** [comment]
please fix this issue

**Comment by ocalhoun6** [comment]
If the draw order is fixed, it might fix #7864 as well.

**Comment by ZehMatt** [comment]
This has nothing to do with draw order, its rather how the top toolbar works and there is no dynamic layout that would shrink the size despite certain buttons missing/being hidden.

**Comment by ocalhoun6** [comment]
You wouldn't need dynamic layout if the rain was drawn first, and then the toolbars were drawn on top of it ... right? The rain could simply be drawn over the entire screen every time, and the toolbar buttons would just cover up the portions behind them.

**Comment by 733737** [comment]
the ui is not drawn on top of anything

**Comment by ZehMatt** [comment]
A little fun fact, the UI is drawn first, there is a "main" window which has the main viewport and that viewport renders all the other windows so no its nothing simple and has also nothing to do with alpha, the rendering process is a pretty complex beast and is not like how ordinary games work, even this basic cheap looking rain is anything far from simple when we are talking about how it works. It will be fixed eventually, all in due time.
